### DISCUSSION TITLE
How can I map the Any key to an actual character.

### DISCUSSION THREAD

**Discussion by cleanshaven** [houmain/keymapper#177] [Q&A] [answered]
In trying to clean up my leftwm key bindings I have been using keymapper.  I'm curious if the following can be done.

Currently I have the following mapping set up.

 ```
   Shift{1} >> sendWindowToTag["1"]
    Shift{2} >> sendWindowToTag["2"]
    Shift{3} >> sendWindowToTag["3"]
    Shift{4} >> sendWindowToTag["4"]
    Shift{5} >> sendWindowToTag["5"]
    Shift{6} >> sendWindowToTag["6"]
    Shift{7} >> sendWindowToTag["7"]
    Shift{8} >> sendWindowToTag["8"]
    Shift{9} >> sendWindowToTag["9"] 

```
I would like to set this up as follows

```
Numbers = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
Shift{Numbers} >> sendWindowToTag[Any]

```
However the Any translates to a key scan code not to a character.

Is there a way to translate the Any key to the actual character

**Comment by houmain** [comment]
This unfortunately won't work since `Any` is evaluated at runtime, while the string to key translation is when the configuration is loaded.
What does your `sendWindowToTag` look like? Maybe these mappings could be generated elegantly with another built-in macro.

**Comment by cleanshaven** [comment]
This is my sendWindowToTag function

`sendWindowToTag = $(leftwm-command "SendWindowToTag $0")`

**Comment by ristomatti** [comment]
Is that same as "Move the focused window to desktop/tag" mentioned here [lefwm keybindings](https://github.com/leftwm/leftwm#default-keys)? If it is, you could then get away by making use of the WM's keybindings:

```
Mod = Meta  # ...change to what you have defined it to be 

Numbers = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
Shift{Numbers} >> Mod{Shift{Numbers}}
```

> However the Any translates to a key scan code not to a character.

Is this what really happens? When I tried this by just making the macro print `$0` to a log file, the only thing that got logged was the string "Any".

**Comment by houmain** [comment]
Yes, right, but I think you meant:

`Shift{Numbers} >> Mod{Shift{Any}}`

> Is this what really happens? When I tried this by just making the macro print $0 to a log file, the only thing that got logged was the string "Any".

This is also what I would have expected.

**Reply by ristomatti** [reply]
@houmain No typos. Isn't it supposed to work? I've used it a lot in my config. Even if it'd mean the same, I find it easier to understand. Examples from [my config](https://github.com/ristomatti/keymapper-config/blob/713538c145173566fd5f536a1c5df41680a58409/keymapper.conf#L751):

```ini
NumpadSymbol       >> NumpadSymbol            # line 89
Hold[Digit]        >> Alt{Digit} ^            # line 751
Hold[FunctionKey]  >> Control{FunctionKey} ^  # line 754
Alt{Direction}     >> (Alt AltGr){Direction}  # line 761
```

**Reply by houmain** [reply]
You are right, this also works, nice.

**Comment by cleanshaven** [comment]
Actually I have disabled leftwm key processing in favor of keymapper.  That is where the leftwm-command program comes in.

I maybe over thinking the problem and can stick to the the current way I am doing it.

Thanks for your help.  I love the way that keymapper works.  I can keep all of my keybindings in one place and move them to multiple systems.

**Reply by ristomatti** [reply]
> I maybe over thinking the problem and can stick to the the current way I am doing it.

Where's the fun in that? You could also "abstract" it away by hiding it behind an `@include`. :thinking:

> Actually I have disabled leftwm key processing in favor of keymapper. That is where the leftwm-command program comes in.

I figured this is the case after reading a bit more about LeftWM. I have to say the developers have come up with a peculiar excuse for rewriting i3wm in Rust. :grin:

**Comment by houmain** [comment]
> >  maybe over thinking the problem and can stick to the the current way I am doing it.

> Where's the fun in that?

There is now a more fun way for doing it. I added a builtin function `apply`, which can generate mappings. The first argument is an expression, which is instantiated with other arguments. So this should work:
```bash
apply[Shift{$0} >> sendWindowToTag[$0], 1, 2, 3, 4, 5, 6, 7, 8, 9]
```

If the expression has multiple arguments, like:
```bash
addMapping = $0 >> $1
apply[addMapping, A, B, C, D]
```
it generates `A >> B` and `C >> D`.

**Comment by cleanshaven** [comment, marked as answer]
This is wonderful.  Just got a chance to get the latest down and compiled.  Thank you for doing this.
